syncapi/storage/shared/syncserver.go

```diff
@@ -912,12 +939,107 @@ func (d *Database) getJoinResponseForCompleteSync(
 	return jr, nil
 }
 
+func (d *Database) getLeaveResponseForCompleteSync(
+	ctx context.Context, txn *sql.Tx,
+	roomID string,
+	r types.Range,
+	filter *gomatrixserverlib.Filter,
+	device userapi.Device,
+) (lr *types.LeaveResponse, err error) {
+	var stateEvents []*gomatrixserverlib.HeaderedEvent
+	stateEvents, err = d.CurrentRoomState.SelectCurrentState(ctx, txn, roomID, &filter.Room.State)
+	if err != nil {
+		return
+	}
+
+	numRecentEventsPerRoom := filter.Room.Timeline.Limit
+
+	// TODO: When filters are added, we may need to call this multiple times to get enough events.
+	//       See: https://github.com/matrix-org/synapse/blob/v0.19.3/synapse/handlers/sync.py#L316
+	var recentStreamEvents []types.StreamEvent
+	var limited bool
+	recentStreamEvents, limited, err = d.OutputEvents.SelectRecentEvents(
+		ctx, txn, roomID, r, numRecentEventsPerRoom, true, true,
+	)
+	if err != nil {
+		return
+	}
+
+	// TODO FIXME: We don't fully implement history visibility yet. To avoid leaking events which the
+	// user shouldn't see, we check the recent events and remove any prior to the join event of the user
+	// which is equiv to history_visibility: joined
+	joinEventIndex := -1
+	leaveEventIndex := -1
+	for i := len(recentStreamEvents) - 1; i >= 0; i-- {
+		ev := recentStreamEvents[i]
+		if ev.Type() == gomatrixserverlib.MRoomMember && ev.StateKeyEquals(device.UserID) {
+			membership, _ := ev.Membership()
+			if membership == gomatrixserverlib.Join {
+				joinEventIndex = i
+				if i > 0 {
+					// the create event happens before the first join, so we should cut it at that point instead
+					if recentStreamEvents[i-1].Type() == gomatrixserverlib.MRoomCreate && recentStreamEvents[i-1].StateKeyEquals("") {
+						joinEventIndex = i - 1
+					}
+				}
+				break
+			} else if membership == gomatrixserverlib.Leave {
+				leaveEventIndex = i
+			}
+
+			if joinEventIndex != -1 && leaveEventIndex != -1 {
+				break
+			}
+		}
+	}
+
+	// We can assume that if the person joined, they also left and will only show the events between (inclusive)
+	if joinEventIndex != -1 && leaveEventIndex != -1 {
+		// cut all events earlier the join (exclude the join itself too)
+		// and cut all events after the person left (exclude the leave itself too)
+		recentStreamEvents = recentStreamEvents[joinEventIndex+1 : leaveEventIndex]
```

Comment:
For reference, this behaves different from [`getJoinResponseForCompleteSync`](https://github.com/matrix-org/dendrite/blob/a85051ca9de82756a94f1ea6870de9bd140a6b20/syncapi/storage/shared/syncserver.go#L888-L889) where we include the `join` event.

In the Sytest for the `leave` response, it expects the first event not to be the join event which is why I have cut it out, [`tests/31sync/09archived.pl#L370-L372`](https://github.com/matrix-org/sytest/blob/11240a80c42994a3de87f6e16b9452dacff2f2c4/tests/31sync/09archived.pl#L370-L372).

But it seems like `join` and `leave` should behave similarly. Which one is right?

Reply:
Discussed this in a sync call with @neilalexander. The join/leave should be included.

The main problem is that there isn't timeline filtering happening when we initially select the events.

Reply:
This was implemented in https://github.com/matrix-org/dendrite/pull/1721